node/hyperbahn/handler.js

```diff
@@ -337,3 +369,128 @@ function unadvertise(service) {
     var self = this;
     self.channel.topChannel.handler.removeServicePeer(service.serviceName, service.hostPort);
 };
+
+function convertHosts(hosts) {
+    var res = [];
+    for (var i = 0; i < hosts.length; i++) {
+        var strs = hosts[i].split(':');
+        var obj = {
+            port: parseInt(strs[1])
+        };
+        strs = strs[0].split('.');
+        obj.ip = {
+            ipv4: parseInt(strs[3]) + (parseInt(strs[2]) << 8) +
+                (parseInt(strs[1]) << 16) + (parseInt(strs[0]) << 24)
+        };
+
+        res.push(obj);
+    }
+
+    return res;
+}
+
+HyperbahnHandler.prototype.discover =
+function discover(self, req, head, body, cb) {
+    var serviceName = body && body.query && body.query.serviceName;
+    if (serviceName.length === 0) {
+        cb(null, {
+            ok: false,
+            body: InvalidServiceName({
+                serviceName: serviceName
+            }),
+            typeName: 'invalidServiceName'
+        });
+        return;
+    }
+
+    var exitNodes = self.egressNodes.exitsFor(serviceName);
+    var exitHosts = Object.keys(exitNodes);
+
+    var myHost = self.channel.hostPort;
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
```

Comment:
Wait; why are we picking a random peer ? 

We can just make a request on the sub channel for the serviceName and sub channel will pick a random peer for us.

We can also skip `waitForIdentified` that way and use retries on other random peers :D